In [1]:
# Read all UCLA datasets.
# There are 5 different datasets.

import csv
import pandas as pd

In [ ]:
# 1. 'usage_bld_kwh.csv' ([geo_id', 'sqft', 'usage', 'usage_med', 'usage_med_sqft', 'usetype', 'year', 'name', 'solar_potential', 'pop', 'usage_percap', 'source'], dtype='object

# 2. 'usage_vintage_kwh.csv' (['geo_id', 'sqft', 'usage', 'usage_med', 'usage_med_sqft', 'usetype', 'vintage_id', 'year', 'name', 'source'], dtype='object')

# 3. 'usage_income_kwh.csv' (['geo_id', 'year', 'income_quintile', 'income_range', 'usage', 'pop', 'usetype', 'usage_percap', 'name', 'source'], dtype='object')

# 4. 'usage_sqft_kwh.csv' (['geo_id', 'sqft', 'sqft_id', 'usage', 'usage_med', 'usage_med_sqft', 'usetype', 'year', 'name', 'source'], dtype='object')

# 5. 'geos.csv' (['geo_id', 'name', 'pop', 'pop_sqmi', 'med_income', 'pct_own', 'pct_rent', 'year', 'enviroscreen_tracts', 'sqmi', 'solar_pot', 'net_solar_pot', 'count', 'source'], dtype='object')



In [2]:
# Dataset 1: Consumption Information by Parcel Use-Type
import pandas as pd

df1 = pd.read_csv('usage_bld_kwh.csv')
df1["source"] = "usage_bld_kwh"
print(df1.columns)

Index(['geo_id', 'sqft', 'usage', 'usage_med', 'usage_med_sqft', 'usetype',
       'year', 'name', 'solar_potential', 'pop', 'usage_percap', 'source'],
      dtype='object')


In [3]:
# Dataset 2: Consumption Information by Building Vintage
import pandas as pd

df2 = pd.read_csv('usage_vintage_kwh.csv')
df2["source"] = "usage_vintage_kwh"
print(df2.columns)

Index(['geo_id', 'sqft', 'usage', 'usage_med', 'usage_med_sqft', 'usetype',
       'vintage_id', 'year', 'name', 'source'],
      dtype='object')


In [7]:
# Dataset 3: Consumption Information by Income Quintile for Residential Accounts 
import pandas as pd

df3 = pd.read_csv('usage_income_kwh.csv')
df3["source"] = "usage_income_kwh"
print(df3.columns)

Index(['geo_id', 'year', 'income_quintile', 'income_range', 'usage', 'pop',
       'usetype', 'usage_percap', 'name', 'source'],
      dtype='object')


In [9]:
# Dataset 4: Consumption information by Square Footage Intervals
import pandas as pd

df4 = pd.read_csv('usage_sqft_kwh.csv')
df4["source"] = "usage_sqft_kwh"
print(df4.columns)

Index(['geo_id', 'sqft', 'sqft_id', 'usage', 'usage_med', 'usage_med_sqft',
       'usetype', 'year', 'name', 'source'],
      dtype='object')


In [11]:
# Dataset 5: Socioeconomic and Solar Information (Add this to the Metadata Word file)
import pandas as pd

df5 = pd.read_csv('geos.csv')
df5["source"] = "geos"
print(df5.columns)

Index(['geo_id', 'name', 'pop', 'pop_sqmi', 'med_income', 'pct_own',
       'pct_rent', 'year', 'enviroscreen_tracts', 'sqmi', 'solar_pot',
       'net_solar_pot', 'count', 'source'],
      dtype='object')


In [13]:
# Merging the Datasets and displaying it

import pandas as pd

df_merged = pd.concat([df1, df2, df3, df4, df5], join='outer').fillna('NA')
# Currently I am doing an outer merge, but I need to double check if I am creating redundancy with
# doing an outer merge. Consider what would happen if we do inner join.

print('The Columns:',df_merged.columns,'\n','-'*100,'\n')
display(df_merged)
ucla_merged = df_merged.copy()

place_names =(ucla_merged['name'].unique()) 
# To see the unique places printed
#for place in place_names:
    #print(place)
#ucla_merged.to_csv('ucla_merged.csv')


The Columns: Index(['geo_id', 'sqft', 'usage', 'usage_med', 'usage_med_sqft', 'usetype',
       'year', 'name', 'solar_potential', 'pop', 'usage_percap', 'source',
       'vintage_id', 'income_quintile', 'income_range', 'sqft_id', 'pop_sqmi',
       'med_income', 'pct_own', 'pct_rent', 'enviroscreen_tracts', 'sqmi',
       'solar_pot', 'net_solar_pot', 'count'],
      dtype='object') 
 ---------------------------------------------------------------------------------------------------- 



,geo_id,sqft,usage,usage_med,usage_med_sqft,usetype,year,name,solar_potential,pop,...,sqft_id,pop_sqmi,med_income,pct_own,pct_rent,enviroscreen_tracts,sqmi,solar_pot,net_solar_pot,count
0,cities_1,1253.0,masked,masked,masked,agriculture,2016,agoura hills,NA,NA,...,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA
1,cities_1,4802115.0,65825608.0000,121901,9.7676,commercial,2016,agoura hills,NA,NA,...,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA
2,cities_1,8906397.0,43707916.6000,9136,3.9839,condo,2016,agoura hills,NA,NA,...,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA
3,cities_1,734698.0,masked,masked,masked,industrial,2016,agoura hills,NA,NA,...,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA
4,cities_1,100192.0,masked,masked,masked,institutional,2016,agoura hills,NA,NA,...,NA,NA,NA,NA,NA,NA,NA,NA,NA,NA
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
957,cities_4,NA,NA,NA,NA,NA,2010,azusa,NA,45301.0,...,NA,3204.573593,59007.0,0.567021,0.432654,4.0,14.136358,255537607.5,-8888.0,-9999.0
958,cities_173,NA,NA,NA,NA,NA,2010,irvine,NA,-7777.0,...,NA,-7777.0,-7777.0,-7777.0,-7777.0,0.0,95.89453,NA,-8888.0,-7777.0
959,cities_182,NA,NA,NA,NA,NA,2010,newport beach,NA,-7777.0,...,NA,-7777.0,-7777.0,-7777.0,-7777.0,0.0,37.307898,NA,-8888.0,-7777.0
960,cities_183,NA,NA,NA,NA,NA,2010,orange,NA,-7777.0,...,NA,-7777.0,-7777.0,-7777.0,-7777.0,1.0,36.512519,NA,-8888.0,-7777.0


In [15]:
# Creating the locations dataframe with columns 'geo_id' and 'name'(location)

location_keys = ucla_merged[['geo_id', 'name']].drop_duplicates().reset_index(drop=True)
#display(location_keys)

# Preparing the Format of Location for Geocoding API
    # Adding ', CA' next to the values of 'name'.
import pandas as pd
locations = location_keys.copy()
locations['query_for_api'] = locations['name'] + ', CA'
display(locations)
#locations.to_csv('locations.csv') No need to save
# Displaying and double checking locations dataframe


,geo_id,name,query_for_api
0,cities_1,agoura hills,"agoura hills, CA"
1,cities_10,carson,"carson, CA"
2,cities_100,camarillo,"camarillo, CA"
3,cities_101,fillmore,"fillmore, CA"
4,cities_102,moorpark,"moorpark, CA"
...,...,...,...
477,neighborhoods_96,calabasas,"calabasas, CA"
478,neighborhoods_97,canoga park,"canoga park, CA"
479,neighborhoods_98,carson,"carson, CA"
480,neighborhoods_99,carthay,"carthay, CA"


In [17]:
# Creating a function to use the Google's Geocoding API

# Structure of the API:
    # https://maps.googleapis.com/maps/api/geocode/json?address=xxx&key=xxx

import requests

def geocode_location(location_name: str, api_key):
    base_url = "https://maps.googleapis.com/maps/api/geocode/json"
    params = {
        "address": location_name,
        "key": api_key
    }
    try:
        response = requests.get(base_url, params=params)
        response.raise_for_status()
        data = response.json()
    except:
        sys.stderr.write(f'Error: The url {base_url} with address {location_name} could not be requested properly!\n')
        return {
            "query": location_name,
            "num_results": 0,
            "data": None,
            "error": "REQUEST_FAILED"
        }
    
    if data['status'] == 'OK':
        len_of_results = len(data['results'])
        result = data['results'][0]
        components = result['address_components']
        print('Data successfully requested.')
        return {"query": location_name, 
                'num_results':len_of_results,
                'data':data
        }
        
    else:
        print(f" Google API error: {data['status']} for '{location_name}'")
        return {
            "query": location_name,
            "num_results": 0,
            "data": None,
            "error": data['status']
        }


In [21]:
# Creating the 'api_results' using the API function I have created

import time

my_api_key = 'PUT YOUR API KEY HERE'
api_results = list()

for index, row in locations.iterrows():
    time.sleep(0.15)
    print(row['geo_id'], ' ', row['query_for_api'])  
    
    response = geocode_location(row['query_for_api'], my_api_key)

    # Append all relevant info
    api_results.append((row['geo_id'], 
                        row['query_for_api'], 
                        {"num_results": response.get("num_results", 0),
                        "error": response.get("error", None),
                        "data": response.get("data", None)} ))


cities_1   agoura hills, CA
Data successfully requested.
cities_10   carson, CA
Data successfully requested.
cities_100   camarillo, CA
Data successfully requested.
cities_101   fillmore, CA
Data successfully requested.
cities_102   moorpark, CA
Data successfully requested.
cities_103   ojai, CA
Data successfully requested.
cities_104   santa paula, CA
Data successfully requested.
cities_105   simi valley, CA
Data successfully requested.
cities_106   thousand oaks, CA
Data successfully requested.
cities_107   port hueneme, CA
Data successfully requested.
cities_108   oxnard, CA
Data successfully requested.
cities_109   san buenaventura, CA
Data successfully requested.
cities_11   cerritos, CA
Data successfully requested.
cities_110   adelanto, CA
Data successfully requested.
cities_111   apple valley, CA
Data successfully requested.
cities_112   barstow, CA
Data successfully requested.
cities_113   big bear lake, CA
Data successfully requested.
cities_114   chino, CA
Data successfully 

In [23]:
# Checking out the first 10 result
for result in api_results[:10]:
    print(result,'\n')

# Turning the api-location results into a dataframe and saving it
import pandas as pd

df_geo = pd.DataFrame(api_results, columns=['geo_id', 'query', 'raw_result'])

# Saving the locations with apis as a csv.
df_geo.to_csv('api_locations.csv')

('cities_1', 'agoura hills, CA', {'num_results': 1, 'error': None, 'data': {'results': [{'address_components': [{'long_name': 'Agoura Hills', 'short_name': 'Agoura Hills', 'types': ['locality', 'political']}, {'long_name': 'Los Angeles County', 'short_name': 'Los Angeles County', 'types': ['administrative_area_level_2', 'political']}, {'long_name': 'California', 'short_name': 'CA', 'types': ['administrative_area_level_1', 'political']}, {'long_name': 'United States', 'short_name': 'US', 'types': ['country', 'political']}, {'long_name': '91301', 'short_name': '91301', 'types': ['postal_code']}], 'formatted_address': 'Agoura Hills, CA 91301, USA', 'geometry': {'bounds': {'northeast': {'lat': 34.168214, 'lng': -118.719776}, 'southwest': {'lat': 34.125329, 'lng': -118.8004199}}, 'location': {'lat': 34.1533395, 'lng': -118.7616764}, 'location_type': 'APPROXIMATE', 'viewport': {'northeast': {'lat': 34.168214, 'lng': -118.719776}, 'southwest': {'lat': 34.125329, 'lng': -118.8004199}}}, 'place

## PROCESSING THE DATA WE GOT USING APIS

**Example Data**: 
<br>{'results': [{'address_components': [{'long_name': 'Agoura Hills', 'short_name': 'Agoura Hills', 'types': ['locality', 'political']},<br>
{'long_name': 'Los Angeles County', 'short_name': 'Los Angeles County', 'types': ['administrative_area_level_2', 'political']}, {'long_name': 'California', 'short_name': 'CA', 'types': ['administrative_area_level_1', 'political']}, {'long_name': 'United States', 'short_name': 'US', 'types': ['country', 'political']},<br> 
{'long_name': '91301', 'short_name': '91301', 'types': ['postal_code']}], 'formatted_address': 'Agoura Hills, CA 91301, USA',<br>
'geometry': {'bounds': {'northeast': {'lat': 34.168214, 'lng': -118.719776}, 'southwest': {'lat': 34.125329, 'lng': -118.8004199}}, 'location': {'lat': 34.1533395, 'lng': -118.7616764}, 'location_type': 'APPROXIMATE', 'viewport': {'northeast': {'lat': 34.168214, 'lng': -118.719776}, 'southwest': {'lat': 34.125329, 'lng': -118.8004199}}}, 'place_id': 'ChIJlRyqIx4h6IAR1v3fLa72Dqw', 'types': ['locality', 'political']}], 'status': 'OK'}

In [1]:
# Reading the 'api_locations' file that I have created using the API
import pandas as pd

api_locations = pd.read_csv('api_locations.csv', index_col=0)

# POST PROCESSING - important
# Making sure that the dictionaries in the dataframe are recognized by the python
import ast
# https://stackoverflow.com/questions/988228/convert-a-string-representation-of-a-dictionary-to-a-dictionary
# I have found this library from stackoverflow, which stated that 'as.literal_eval' as a safer alternative than just using 'eval' library.

print(isinstance(api_locations['raw_result'].iloc[0], dict))

# I need a double check for this because this gives an error when the code already is a dictionary.
if isinstance(api_locations['raw_result'].iloc[0], str):
    api_locations['raw_result'] = api_locations['raw_result'].apply(ast.literal_eval)

print(isinstance(api_locations['raw_result'].iloc[0], dict))
display(api_locations)


False
True


,geo_id,query,raw_result
0,cities_1,"agoura hills, CA","{'num_results': 1, 'error': None, 'data': {'re..."
1,cities_10,"carson, CA","{'num_results': 1, 'error': None, 'data': {'re..."
2,cities_100,"camarillo, CA","{'num_results': 1, 'error': None, 'data': {'re..."
3,cities_101,"fillmore, CA","{'num_results': 1, 'error': None, 'data': {'re..."
4,cities_102,"moorpark, CA","{'num_results': 1, 'error': None, 'data': {'re..."
...,...,...,...
477,neighborhoods_96,"calabasas, CA","{'num_results': 1, 'error': None, 'data': {'re..."
478,neighborhoods_97,"canoga park, CA","{'num_results': 1, 'error': None, 'data': {'re..."
479,neighborhoods_98,"carson, CA","{'num_results': 1, 'error': None, 'data': {'re..."
480,neighborhoods_99,"carthay, CA","{'num_results': 1, 'error': None, 'data': {'re..."


In [3]:
# Creating a function that extracts the relevant information from the dictionary in 'raw result'
def extract_location_components(raw_result):
    try:
        first_result = raw_result['data']['results'][0] # Using the first capture
        components = first_result['address_components']
        geometry = first_result['geometry']['location']

        # Initialize output
        output = {
            "postal_code": None,
            "city": None,
            "county": None,
            "neighborhood": None,
            "lat": geometry.get('lat'),
            "lng": geometry.get('lng')
        }

        for comp in components:
            types = comp['types']
            if 'postal_code' in types:
                output['postal_code'] = comp['long_name']
            if 'locality' in types:
                output['city'] = comp['long_name']
            if 'administrative_area_level_2' in types:
                output['county'] = comp['long_name']
            if 'neighborhood' in types:
                output['neighborhood'] = comp['long_name']

        return output
    except:
        return {
            "postal_code": None,
            "city": None,
            "county": None,
            "neighborhood": None,
            "lat": None,
            "lng": None
        }


In [5]:
# Applying this function to every row to generate the 'location details' column
api_locations['location_details'] = api_locations['raw_result'].apply(extract_location_components)
display(api_locations)
# I do have 271 neighborhood information


,geo_id,query,raw_result,location_details
0,cities_1,"agoura hills, CA","{'num_results': 1, 'error': None, 'data': {'re...","{'postal_code': '91301', 'city': 'Agoura Hills..."
1,cities_10,"carson, CA","{'num_results': 1, 'error': None, 'data': {'re...","{'postal_code': None, 'city': 'Carson', 'count..."
2,cities_100,"camarillo, CA","{'num_results': 1, 'error': None, 'data': {'re...","{'postal_code': None, 'city': 'Camarillo', 'co..."
3,cities_101,"fillmore, CA","{'num_results': 1, 'error': None, 'data': {'re...","{'postal_code': '93015', 'city': 'Fillmore', '..."
4,cities_102,"moorpark, CA","{'num_results': 1, 'error': None, 'data': {'re...","{'postal_code': '93021', 'city': 'Moorpark', '..."
...,...,...,...,...
477,neighborhoods_96,"calabasas, CA","{'num_results': 1, 'error': None, 'data': {'re...","{'postal_code': None, 'city': 'Calabasas', 'co..."
478,neighborhoods_97,"canoga park, CA","{'num_results': 1, 'error': None, 'data': {'re...","{'postal_code': None, 'city': 'Los Angeles', '..."
479,neighborhoods_98,"carson, CA","{'num_results': 1, 'error': None, 'data': {'re...","{'postal_code': None, 'city': 'Carson', 'count..."
480,neighborhoods_99,"carthay, CA","{'num_results': 1, 'error': None, 'data': {'re...","{'postal_code': '90035', 'city': 'Los Angeles'..."


------------------------------------------------------------------------------------------------

In [8]:
# Expand the new informations in 'location_details' into new columns

location_df = pd.DataFrame(api_locations['location_details'].tolist())
#display(location_df)
api_locations_processed = pd.concat([api_locations, location_df], axis=1)

display(api_locations_processed)


,geo_id,query,raw_result,location_details,postal_code,city,county,neighborhood,lat,lng
0,cities_1,"agoura hills, CA","{'num_results': 1, 'error': None, 'data': {'re...","{'postal_code': '91301', 'city': 'Agoura Hills...",91301,Agoura Hills,Los Angeles County,None,34.153340,-118.761676
1,cities_10,"carson, CA","{'num_results': 1, 'error': None, 'data': {'re...","{'postal_code': None, 'city': 'Carson', 'count...",None,Carson,Los Angeles County,None,33.831674,-118.281693
2,cities_100,"camarillo, CA","{'num_results': 1, 'error': None, 'data': {'re...","{'postal_code': None, 'city': 'Camarillo', 'co...",None,Camarillo,Ventura County,None,34.216394,-119.037602
3,cities_101,"fillmore, CA","{'num_results': 1, 'error': None, 'data': {'re...","{'postal_code': '93015', 'city': 'Fillmore', '...",93015,Fillmore,Ventura County,None,34.399164,-118.918154
4,cities_102,"moorpark, CA","{'num_results': 1, 'error': None, 'data': {'re...","{'postal_code': '93021', 'city': 'Moorpark', '...",93021,Moorpark,Ventura County,None,34.285558,-118.882041
...,...,...,...,...,...,...,...,...,...,...
477,neighborhoods_96,"calabasas, CA","{'num_results': 1, 'error': None, 'data': {'re...","{'postal_code': None, 'city': 'Calabasas', 'co...",None,Calabasas,Los Angeles County,None,34.153973,-118.647380
478,neighborhoods_97,"canoga park, CA","{'num_results': 1, 'error': None, 'data': {'re...","{'postal_code': None, 'city': 'Los Angeles', '...",None,Los Angeles,Los Angeles County,Canoga Park,34.208309,-118.597532
479,neighborhoods_98,"carson, CA","{'num_results': 1, 'error': None, 'data': {'re...","{'postal_code': None, 'city': 'Carson', 'count...",None,Carson,Los Angeles County,None,33.831674,-118.281693
480,neighborhoods_99,"carthay, CA","{'num_results': 1, 'error': None, 'data': {'re...","{'postal_code': '90035', 'city': 'Los Angeles'...",90035,Los Angeles,Los Angeles County,Carthay Square,34.055838,-118.368030


In [10]:
# Read the 'api_locations_processed' dataframe
import csv
import pandas as pd

#api_loc_process = pd.read_csv('api_locations_processed.csv', dtype={'postal_code': str})
api_loc_process = api_locations_processed.copy()

# Dropping the index columns I don't need them
api_loc_process = api_loc_process.loc[:, ~api_loc_process.columns.str.startswith('Unnamed:')]

display(api_loc_process)

##################################################################################
#482 unique location combinations.

# 482x100x0.8 = 38560 house entries.
# Let's assume some will be duplicates and some will note be useful for me:
# I will have 30k houses for sure.

#38962 rows from the UCLA data, with extra columns derived using the API.
# 5,773 rows from usage_bld_kwh
# 23,089 rows from usage_vintage_kwh
# 2,406 usage_income_kwh.csv
# 6,735 rows from usage_sqft_kwh.csv
# 963 rows from geos.csv

# histogram of locations
# use the most occuring 150 locations.


,geo_id,query,raw_result,location_details,postal_code,city,county,neighborhood,lat,lng
0,cities_1,"agoura hills, CA","{'num_results': 1, 'error': None, 'data': {'re...","{'postal_code': '91301', 'city': 'Agoura Hills...",91301,Agoura Hills,Los Angeles County,None,34.153340,-118.761676
1,cities_10,"carson, CA","{'num_results': 1, 'error': None, 'data': {'re...","{'postal_code': None, 'city': 'Carson', 'count...",None,Carson,Los Angeles County,None,33.831674,-118.281693
2,cities_100,"camarillo, CA","{'num_results': 1, 'error': None, 'data': {'re...","{'postal_code': None, 'city': 'Camarillo', 'co...",None,Camarillo,Ventura County,None,34.216394,-119.037602
3,cities_101,"fillmore, CA","{'num_results': 1, 'error': None, 'data': {'re...","{'postal_code': '93015', 'city': 'Fillmore', '...",93015,Fillmore,Ventura County,None,34.399164,-118.918154
4,cities_102,"moorpark, CA","{'num_results': 1, 'error': None, 'data': {'re...","{'postal_code': '93021', 'city': 'Moorpark', '...",93021,Moorpark,Ventura County,None,34.285558,-118.882041
...,...,...,...,...,...,...,...,...,...,...
477,neighborhoods_96,"calabasas, CA","{'num_results': 1, 'error': None, 'data': {'re...","{'postal_code': None, 'city': 'Calabasas', 'co...",None,Calabasas,Los Angeles County,None,34.153973,-118.647380
478,neighborhoods_97,"canoga park, CA","{'num_results': 1, 'error': None, 'data': {'re...","{'postal_code': None, 'city': 'Los Angeles', '...",None,Los Angeles,Los Angeles County,Canoga Park,34.208309,-118.597532
479,neighborhoods_98,"carson, CA","{'num_results': 1, 'error': None, 'data': {'re...","{'postal_code': None, 'city': 'Carson', 'count...",None,Carson,Los Angeles County,None,33.831674,-118.281693
480,neighborhoods_99,"carthay, CA","{'num_results': 1, 'error': None, 'data': {'re...","{'postal_code': '90035', 'city': 'Los Angeles'...",90035,Los Angeles,Los Angeles County,Carthay Square,34.055838,-118.368030


In [14]:
# Next step is to properly merge the 'api_loc_process' with the 'ucla_merged.csv' file.

import pandas as pd

# Load data
ucla_df = pd.read_csv('ucla_merged.csv')
api_df = api_loc_process.copy()

# Merge on geo_id (left join to preserve all ucla_df rows)
merged_df = pd.merge(ucla_df, api_df, on='geo_id', how='left', suffixes=('', '_api'))

# Keep only rows where 'name' is found in 'query' (or query is NaN so that ucla rows are retained)
merged_df = merged_df[merged_df['query'].isna() | merged_df.apply(lambda row: row['name'] in row['query'], axis=1)]

# Drop extra index columns if present
merged_df = merged_df.loc[:, ~merged_df.columns.str.startswith('Unnamed:')]

# Save the file with every row
##merged_df.to_csv('ucla_merged_with_api.csv', index=False)

# Save the final deliverable with trimming unnecessdary columns out.
columns_to_keep = ['geo_id','sqft','usage','usage_med','usage_med_sqft','usetype',
                   'year','name','solar_potential','pop','usage_percap','source',
                   'vintage_id','income_quintile','income_range','sqft_id','pop_sqmi',
                   'med_income','pct_own','pct_rent','enviroscreen_tracts','sqmi',
                  'solar_pot','net_solar_pot','count','query','postal_code','city','county','neighborhood','lat','lng']

final_df_col = merged_df[columns_to_keep]

# Save the final file
final_df_col.to_csv('ucla_final.csv', index=False)

# Saving the final file into Craigslist_Scraping
final_df_col.to_csv('../Craigslist_Scraping/ucla_final.csv', index=False)

# Saving the final file into Predicting_Energy
final_df_col.to_csv('../Predicting_Energy/ucla_final.csv', index=False)


/var/folders/w_/_pjdhc9x5pndg2r8dn0lh6h00000gn/T/ipykernel_39718/3295976022.py:6: DtypeWarning: Columns (11,15) have mixed types. Specify dtype option on import or set low_memory=False.
  ucla_df = pd.read_csv('ucla_merged.csv')
